ddtrace/sampler.py

```diff
@@ -92,3 +103,251 @@ def set_sample_rate_by_service(self, rate_by_service):
 
 # Default key for service with no specific rate
 RateByServiceSampler._default_key = RateByServiceSampler._key()
+
+
+class DatadogSampler(BaseSampler):
+    """
+    """
+    __slots__ = ('rules', 'rate_limit')
+
+    DEFAULT_RATE_LIMIT = 100
+    NO_RATE_LIMIT = -1
+
+    def __init__(self, rules=None, rate_limit=None):
+        """
+        Constructor for DatadogSampler sampler
+
+        :param rules: List of :class:`SamplingRule` rules to apply to the root span of every trace, default no rules
+        :type rules: :obj:`list` of :class:`SamplingRule`
+        :param rate_limit: Global rate limit (traces per second) to apply to all traces regardless of the rules
+            applied to them, default 100 traces per second
+        :type rate_limit: :obj:`int`
+        """
+        # Ensure rules is a list
+        if not rules:
+            rules = []
+
+        # Validate that the rules is a list of SampleRules
+        for rule in rules:
+            if not isinstance(rule, SamplingRule):
+                raise TypeError('Rule {!r} must be a sub-class of type ddtrace.sampler.SamplingRules'.format(rule))
+        self.rules = rules
+
+        # Configure rate limiter
+        if rate_limit is None:
+            rate_limit = self.DEFAULT_RATE_LIMIT
+        self.limiter = RateLimiter(rate_limit)
+
+    def sample(self, span):
+        """
+        Decide whether the provided span should be sampled or not
+
+        The span provided should be the root span in the trace.
+
+        :param span: The root span of a trace
+        :type span: :class:`ddtrace.span.Span`
+        :returns: Whether the span was sampled or not
+        :rtype: :obj:`bool`
+        """
+        # If there are rules defined, then iterate through them and find one that wants to sample
+        matching_rule = None
+        if self.rules:
+            # Go through all rules and grab the first one that matched
+            # DEV: This means rules should be ordered by the user from most specific to least specific
+            for rule in self.rules:
+                if rule.matches(span):
+                    matching_rule = rule
+                    break
+            else:
+                # No rule matched, reject
```

Comment:
According to the RFC, no rule matched should fallback to the agent rate. This is to allow customers to enable the sampler for example on only one resource without a huge regression on the rest of the traces.

Reply:
I disagree with that. We should not fallback to old behavior, if we are using this sampler this should be the only logic in use.

The fallback mechanism adds way more complexity to this implementation and should be avoided.

Reply:
One of the use cases is that as a customer I should be able to use the new sampling logic only on a specific service and/or span name and/or resource. How can this be achieved with the current code without breaking the smart sampling for unmatched spans?

Reply:
I think there are a few cases to consider (when no rules are put into place):

1. A customer generating < 100 traces per second per tracer.
  a. They will have everything sampled as a p1, no issues
  b. It also doesn't matter how many services they are reporting per tracer since they are under the limit anyways
2. A customer generating > 100 traces per second per tracer, but only 1 top level service per tracer
  a. It won't necessarily be "uniform" in the correct sense since it'll just mark everything over 100 traces per second a p0
  b. Signature sampling in the agent should pick up on some interesting p0s being sent
3. A customer with multiple top level services per tracer sending > 100 traces per second
  a. This case isn't ideal since we do not get uniform sampling across services with regards to the 100 traces per second limit
  b. If Service A generates 100 traces, and then Service B generates 100 all within a second, then Service B will never have any traces sampled, no bueno
  c. Signature sampling in the agent should pick up a few interesting p0 traces
4. (Are there others I forgot? These seem to be the most common cases I can think of)

I think condition 1 is likely the most common across all of our customers, and most people are probably fine (with the new 100 traces per second per tracer limit in place).

I think conditions 2 and 3 are mostly the cases with our larger customers, condition 2 is probably more common than condition 3.

For condition 2, we give people the ability with this new sampler to limit how many traces are generated for certain resources so they have the control to balance their traces amongst the 100 traces per second.

For condition 3, we give people the ability with this new sampler to limit how many traces are generates for a certain service so they have the control to balance their traces amongst the 100 traces per second.

(also everyone has the ability to bump the 100 traces per second up to a higher number as well).

While we can do better to not force people to manually add rules to control their sampling, the benefits of this new sampler is to be easy to understand and easy to control on your own.

Reply:
One main advantage I see of the fallback to the RateByService is that the switch to the `DatadogSampler` isn't a "breaking change" beyond just the custom rules added. If you plug it with no rule (or only a very scoped rule which is what we'd expect from most common usage), you should basically end up with the same behavior as before. Here you would basically disable priority sampling, breaking existing distributed traces.

Also, the fallback to the RateByService is only "temporarily". Later that'd be replaced by a "smart sampling" taking smarter default decision.

Reply:
> Here you would basically disable priority sampling, breaking existing distributed traces.

This should not be the case at all, since sampling priority is distributed and overridden from headers/metadata anyways. A p1 trace should be a p1 trace on every tracer it hits along it's way.

> Also, the fallback to the RateByService is only "temporarily". Later that'd be replaced by a "smart sampling" taking smarter default decision.

We should be discussing this logic and a timeline for execution as well. If it only adds 2 weeks of work (making a generalized statement), then we should prioritize that work and just release everything at once. 

I want to avoid mixed behaviors as much as possible. Our sampling logic is already too difficult to manage, every opportunity we have to make it simpler will be better for us and for users.

> Anyway, if we really think that the RFC is incorrect, let's be sure the discussion is happening (or at least referred) there.

We want to try and complete the prototype and get some real usage of it. Rather than discuss theoretical on what will/won't be best, let's try it internally and see how it actually works/what changes are required.